pkg/splunk/enterprise/afwscheduler.go

```diff
@@ -781,12 +776,191 @@ podCopyPhase:
+	podID, err := getOrdinalValFromPodName(podName)
+	if err != nil {
+		scopedLog.Error(err, "unable to derive podId for podname", podName)
+		return false
+	}
+
+	select {
+	case installTracker[podID] <- struct{}{}:
+		return true
+	default:
+		return false
+	}
+}
+
+// freeInstallSlotForPod frees up an install slot for a pod
+func freeInstallSlotForPod(installTracker []chan struct{}, podName string) {
+	scopedLog := log.WithName("freeInstallSlotForPod")
+	podID, err := getOrdinalValFromPodName(podName)
+	if err != nil {
+		scopedLog.Error(err, "unable to derive podId for podname", podName)
+		return
+	}
+
+	select {
+	case <-installTracker[podID]:
+	default:
+		scopedLog.Error(nil, "trying to free an install slot without even allocating it")
+	}
+}
+
+// isPendingClusterScopeWork confirms if there is any pending custer scoped app work
+func isPendingClusterScopeWork(afwPipeline *AppInstallPipeline) bool {
+	// CR doesn't deal with cluster scoped apps
+	if !isClusterScoped(afwPipeline.cr.GetObjectKind().GroupVersionKind().Kind) {
+		return false
+	}
+
+	// There is no cluster scoped apps pending for bundle push
+	if afwPipeline.appDeployContext.BundlePushStatus.BundlePushStage == enterpriseApi.BundlePushComplete || afwPipeline.appDeployContext.BundlePushStatus.BundlePushStage == enterpriseApi.BundlePushUninitialized {
+		return false
+	}
+
+	return true
+}
+
+// needToRunClusterScopedPlaybook confirms if the cluster scoped playbooks to be run
+func needToRunClusterScopedPlaybook(afwPipeline *AppInstallPipeline) bool {
+	if !isPendingClusterScopeWork(afwPipeline) {
+		return false
+	}
+
+	// Its already time to yield the current reconcile
+	if afwPipeline.afwEntryTime+maxRunTimeBeforeAttemptingYield < time.Now().Unix() {
+		return false
+	}
+
+	return true
+}
+
+// tryAppPkgCleanupFromOperatorPod tries to change the app install status, also cleans the app pkg from Operator Pod
+func tryAppPkgCleanupFromOperatorPod(installWorker *PipelineWorker) {
+	scopedLog := log.WithName("tryAppPkgCleanupFromOperatorPod")
+	if installWorker.sts == nil {
+		scopedLog.Error(nil, "sts is missing", "cr", installWorker.cr.GetName(), "kind", installWorker.cr.GroupVersionKind().Kind, "app pkg", installWorker.appDeployInfo.AppName)
+		deleteAppPkgFromOperator(installWorker)
+		return
+	}
+
+	if *installWorker.sts.Spec.Replicas > 1 {
+		if isAppInstallationCompleteOnAllReplicas(installWorker.appDeployInfo.AuxPhaseInfo) {
+			scopedLog.Info("app pkg installed on all the pods", "app pkg", installWorker.appDeployInfo.AppName)
+			installWorker.appDeployInfo.PhaseInfo.Phase = enterpriseApi.PhaseInstall
+			installWorker.appDeployInfo.PhaseInfo.Status = enterpriseApi.AppPkgInstallComplete
+
+			//For now, set the deploy status as complete. Eventually, we can phase it out
+			installWorker.appDeployInfo.DeployStatus = enterpriseApi.DeployStatusComplete
+			deleteAppPkgFromOperator(installWorker)
+		}
+	} else {
+		deleteAppPkgFromOperator(installWorker)
+	}
```

Comment:
Probably another way to write this is -
```
if *installWorker.sts.Spec.Replicas > 1 &&  isAppInstallationCompleteOnAllReplicas(installWorker.appDeployInfo.AuxPhaseInfo) {
    scopedLog.Info("app pkg installed on all the pods", "app pkg", installWorker.appDeployInfo.AppName)
    installWorker.appDeployInfo.PhaseInfo.Phase = enterpriseApi.PhaseInstall
    installWorker.appDeployInfo.PhaseInfo.Status = enterpriseApi.AppPkgInstallComplete

    //For now, set the deploy status as complete. Eventually, we can phase it out
    installWorker.appDeployInfo.DeployStatus = enterpriseApi.DeployStatusComplete
}
deleteAppPkgFromOperator(installWorker)
```

Reply:
TBH, that is where i started :-), then noticed,  it can pre-maturely delete the app package before it was copied to all the Pods, once the first install is complete on a replica member.

Reply:
Oh how is that? Shouldn't `isAppInstallationCompleteOnAllReplicas` take care of that?

Reply:
For example, say, if the replica count is 2, and when the first installation is complete, then we enter this block of code. So, in that case, as `isAppInstallationCompleteOnAllReplicas ()` returns false, we end up deleting the package, though the installation is not complete on all the pods. Hope this clarifies :-)